sw/airborne/modules/ctc/ctc.c

```diff
@@ -0,0 +1,285 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ */
+
+#include <math.h>
+#include <std.h>
+#include <stdio.h>
+
+#include "modules/ctc/ctc.h"
+#include "subsystems/datalink/datalink.h" // dl_buffer
+#include "subsystems/datalink/telemetry.h"
+#include "subsystems/navigation/common_nav.h"
+#include "firmwares/fixedwing/stabilization/stabilization_attitude.h"
+#include "autopilot.h"
+
+/*! Max expected number of aircraft */
+#ifndef CTC_MAX_AC
+#define CTC_MAX_AC 4
+#endif
+
+#if PERIODIC_TELEMETRY
+static void send_ctc(struct transport_tx *trans, struct link_device *dev)
+{
+  pprz_msg_send_CTC(trans, dev, AC_ID, 6 * CTC_MAX_AC, &(tableNei[0][0]));
+}
+
+static void send_ctc_control(struct transport_tx *trans, struct link_device *dev)
+{
+  pprz_msg_send_CTC_CONTROL(trans, dev, AC_ID, &ctc_control.v_centroid_x, &ctc_control.v_centroid_y,
+                            &ctc_control.target_vx, &ctc_control.target_vy, &ctc_control.target_px, &ctc_control.target_py,
+                            &ctc_control.ref_px, &ctc_control.ref_py);
+}
+#endif // PERIODIC TELEMETRY
+
+// Control
+/*! Default gains for the algorithm */
+#ifndef CTC_GAIN_K1
+#define CTC_GAIN_K1 0.001
+#endif
+#ifndef CTC_GAIN_K2
+#define CTC_GAIN_K2 0.001
+#endif
+#ifndef CTC_GAIN_ALPHA
+#define CTC_GAIN_ALPHA 0.01
+#endif
+/*! Default timeout in ms for the neighbors' information */
+#ifndef CTC_TIMEOUT
+#define CTC_TIMEOUT 1500
+#endif
+/*! Default angular velocity around target in rad/sec */
+#ifndef CTC_OMEGA
+#define CTC_OMEGA 0.25
+#endif
+/*! Default time in ms for broadcasting information */
+#ifndef CTC_TIME_BROAD
+#define CTC_TIME_BROAD 100
+#endif
+
+ctc_con ctc_control = {CTC_GAIN_K1, CTC_GAIN_K2, CTC_GAIN_ALPHA, CTC_TIMEOUT,
+                       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, CTC_OMEGA, CTC_TIME_BROAD
+                      };
+
+int16_t tableNei[CTC_MAX_AC][6];
+uint32_t last_info[CTC_MAX_AC];
+uint32_t last_transmision = 0;
+uint32_t before = 0;
+
+uint32_t time_init_table;
+
+float moving_target_px = 0;
+float moving_target_py = 0;
+float moving_target_vx = 0;
+float moving_target_vy = 0;
+
+void ctc_init(void)
+{
+  for (int i = 0; i < CTC_MAX_AC; i++) {
+    tableNei[i][0] = -1;
+  }
+
+#if PERIODIC_TELEMETRY
+  register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_CTC, send_ctc);
+  register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_CTC_CONTROL, send_ctc_control);
+#endif
+}
+
+bool collective_tracking_vehicle()
+{
+    ctc_control.target_px = moving_target_px;
+    ctc_control.target_py = moving_target_py;
+    ctc_control.target_vx = moving_target_vx;
+    ctc_control.target_vy = moving_target_vy;
+
+    collective_tracking_control();
+
+    return true;
+}
+
+bool collective_tracking_waypoint(uint8_t wp)
```

Comment:
`wp` is not used here ? same for `x,y` below ?

Reply:
fixed